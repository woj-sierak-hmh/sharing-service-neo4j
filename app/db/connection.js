import { v1 as neo4j } from 'neo4j-driver';
import config from '../config.js';

let driverSession = null;

// export default driver.session();
export const getSession = () => {
  if (!driverSession) {
    const driver = neo4j.driver(
      config.get('DB_URI'),
      neo4j.auth.basic(
        config.get('DB_AUTHB_USERNAME'),
        config.get('DB_AUTHB_PASSWORD')
      )
    );
    driverSession = driver.session();
  }
  return driverSession;
};
