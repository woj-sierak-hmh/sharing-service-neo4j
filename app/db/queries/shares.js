import session from '../connection.js';

export const createShare = async ({
  tenantRefId,
  sharerRefId,
  assetType,
  assetRefId,
  recipients,
}) => {
  const result = await session.run(
    'MERGE (org:Organization {orgRefId: {tenantRefId}}) ' +
      'MERGE (asset:Asset {assetRefId: {assetRefId}, assetType: {assetType}, isActive: "true"}) ' +
      'MERGE (org)-[:MASTER_OF]->(asset) ' +
      'MERGE (creator:User {userRefId: {sharerRefId}}) ' +
      'MERGE (creator)-[:CREATOR_OF]->(asset) ' +
      'FOREACH (ref IN {recipients} | ' +
      '  MERGE (recipient:User {userRefId: ref}) ' +
      '  MERGE (asset)-[:SHARED_WITH {type:"READ", createdDate: $createdDate}]->(recipient))',
    {
      tenantRefId,
      sharerRefId,
      assetType,
      assetRefId,
      createdDate: new Date().toISOString(),
      // temporarily...
      // recipient: `teacher` + Math.ceil(Math.random() * 100)
      recipients,
    }
  );
  session.close();
  return result;
};

export const getShares = async ({ tenantRefId, userRefId, assetType }) => {
  const results = await session.run(
    'MATCH (:User {userRefId: {userRefId}})<-[share:SHARED_WITH]-(asset:Asset {assetType: {assetType}})<-[:MASTER_OF]-(:Organization {orgRefId: {tenantRefId}}) ' +
      'WITH share, asset ' +
      'MATCH (asset)<-[:CREATOR_OF]-(owner:User) ' +
      'RETURN asset.assetRefId AS assetRefId, asset.assetType AS assetType, ' +
      'share.createdDate AS shareDate, owner.userRefId AS sharerRefId ' +
      'ORDER BY shareDate DESC',
    {
      tenantRefId,
      userRefId,
      assetType,
    }
  );
  const assets = results.records.map(r => {
    return {
      assetRefId: r.get('assetRefId'),
      assetType: r.get('assetType'),
      shareRefId: r.get('sharerRefId'),
      shareDate: r.get('shareDate'),
    };
  });
  return assets;
};
