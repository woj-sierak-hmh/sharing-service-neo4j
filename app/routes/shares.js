import * as shares from '../db/queries/shares.js';

export const createShare = async (ctx, next) => {
  // TODO validation of params

  ctx.log.info('ctx.request.body:', ctx.request.body);
  const recipients = ctx.request.body.recipientRefIds;
  ctx.log.info('recipientRefIds:', recipients);

  try {
    const result = await shares.createShare({
      ...ctx.params,
      recipients: recipients || ['grzuby'],
    });
    ctx.log.info('--------------->', result.summary.statement.parameters);
    ctx.status = 202;
  } catch (err) {
    ctx.log.error(err);
    ctx.status = 500;
    ctx.body = 'Apparently something went wrong...' + err.code;
  }

  // on application exit:
  // driver.close();
};

export const getShares = async ctx => {
  ctx.log.info('params-->', ctx.params);
  try {
    const assets = await shares.getShares(ctx.params);
    ctx.status = 200;
    ctx.body = { assets };
  } catch (err) {
    ctx.log.error(err);
    ctx.status = 500;
    ctx.body = 'Apparently something went wrong...' + err.code;
  }
};
