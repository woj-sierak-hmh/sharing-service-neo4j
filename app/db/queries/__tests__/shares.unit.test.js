import { getSession } from '../../connection.js';
import { createShare } from '../shares.js';

const runRet = { foo: 'bar' };
const mockRun = jest.fn(() => runRet);
const mockClose = jest.fn();

jest.mock('../../connection.js', () => {
  return {
    getSession: jest.fn().mockImplementation(() => {
      return {
        run: mockRun,
        close: mockClose,
      };
    }),
  };
});

describe('shares', () => {
  beforeAll(() => {
    getSession.mockClear();
    // getSession.mockImplementation(() => {
    //   return {
    //     run: mockRun,
    //     close: mockClose,
    //   };
    // });
    //console.log(getSession);
  });
  afterEach(() => {
    // getSession.mockClear();
  });
  test('createShare', async () => {
    const inputObj = {
      tenantRefId: 'tenantRefId',
      sharerRefId: 'sharerRefId',
      assetType: 'assetType',
      assetRefId: 'assetRefId',
      recipients: 'recipients',
    };

    const res = await createShare(inputObj);
    console.log('--->', getSession.mock.calls);

    expect(mockRun).toBeCalled();
    console.log('aa->', mockRun.mock.calls[0]);
    expect(mockRun.mock.calls[0][0]).toMatchSnapshot();
    expect(mockRun.mock.calls[0][1]).toMatchSnapshot({
      createdDate: expect.any(String),
    });

    expect(mockClose).toBeCalled();
    expect(res).toBe(runRet);

    // expect(res).toBe('-done-');
    // expect(run).toHaveBeenCalled(); // With(inputObj);
    // expect(close).toHaveBeenCalled();
  });
});
