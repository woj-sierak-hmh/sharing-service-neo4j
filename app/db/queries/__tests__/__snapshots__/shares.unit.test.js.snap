// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shares createShare 1`] = `"MERGE (org:Organization {orgRefId: {tenantRefId}}) MERGE (asset:Asset {assetRefId: {assetRefId}, assetType: {assetType}, isActive: \\"true\\"}) MERGE (org)-[:MASTER_OF]->(asset) MERGE (creator:User {userRefId: {sharerRefId}}) MERGE (creator)-[:CREATOR_OF]->(asset) FOREACH (ref IN {recipients} |   MERGE (recipient:User {userRefId: ref})   MERGE (asset)-[:SHARED_WITH {type:\\"READ\\", createdDate: $createdDate}]->(recipient))"`;

exports[`shares createShare 2`] = `
Object {
  "assetRefId": "assetRefId",
  "assetType": "assetType",
  "createdDate": Any<String>,
  "recipients": "recipients",
  "sharerRefId": "sharerRefId",
  "tenantRefId": "tenantRefId",
}
`;
